### ISSUE TITLE
Combat Bug / fight in progress after wipe (15$ bounty)

### ISSUE THREAD

**Issue by durrrr** [TrinityCore/TrinityCore#6756]
Hello,
I have encountered a problem while playing in some game zones, such as "Naxxaramas", "Trial of the Crusader" and more.

How to reproduce the bug:
1. Finish some of the encounters and stay in combat mode.

What should happen, what happens instead:
You should finish the encounters and get out of Combat mode, what happens instead is you get stuck in combat mode with no ability to get out of combat unless you use cd's like "Vanish".

This issue is problematic in some places because it holds you back from summoning, getting loots, and more.
Achievements like "The Immortal" are nearly impossible because if someone dc's or relogs he just dies, and then the achievement is just ruined by script.

Some of the NPC's that keeps you in combat bug:
http://www.wowhead.com/npc=15932 --> The NPC #1 (Gluth) (Old Wowhead)
http://www.wowhead.com/npc=15928 --> The NPC #2 (Thaddius) (Old Wowhead)
http://www.wowhead.com/npc=34564 --> The NPC #3 (Anub'arak) (Old Wowhead)
http://www.wowhead.com/npc=31139 --> The NPC #4 (Pustulent Horror) (Old Wowhead)
http://www.wowhead.com/npc=30330 --> The NPC #5 (Jotunheim Proto-Drake) (Old Wowhead)

These 3 are just a small example. I cant remember them all so i'll be glad of some other people will implement their NPC's which keeps them in combat.

I'm running Unix on hash: 7e62c34 , TDB 335.11.46+
Addons: AH bot and WG.

Regards,
durrrr
##

**Comment by Runico** [comment]
this still happens

**Comment by sirikfoll** [comment]
Instance stuck in combat because of Mimiron's fight.
How to reproduce: with two players, leave player1 inside the instance, like at the entrance. 
Player2 start the fight and die(or set gm on) as soon as the boss mount the vehicle. The boss will remain with aggro with the player1. This can also be done during the phase transitions

Did some debugging and this is what i realized
Using the reproduction method above, it can be separate in 2 different cases, if the vehicle is or not in combat by the time player2 die.(If player2 die immediately after starting the fight, the vehicle will not be in combat).

1. If the vehicle isn't in combat, then player1 can leave the instance without reseting the boss. 
Why?
Because for mimiron, UpdateVictim() returns true because he has React State Passive, so it only checks if there's someone in his ThreatList.
And if the vehicle is not in combat, it won't check anything nor evade.

2. If the vehicle is in combat, it will remain stuck in combat only if player1 remains  inside the instance.
Why?
For the vehicles, Mkii for example, UpdateVictim() will return true, because of this check
https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Entities/Unit/Unit.cpp#L12774 (all three checks here return true in these places below)         https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Entities/Creature/Creature.cpp#L2082 (possible the problem is here, it return true before check the distance)
         https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/server/game/Entities/Creature/Creature.cpp#L2048

**Comment by kamicolo** [comment]
/cough -> critters seriously go and try reproduce the issue without any critters inside the DB.

**Comment by Aokromes** [comment]
added another 15$ bounty here.

**Comment by xinef1** [comment]
This will solve worldwide combat bugs. It basicly removes all hostile references that are out of our sight (the cause of combat bug). As for instance scripts, they need to be fixed separately.

<details>

```diff
 src/server/game/Entities/Player/Player.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/src/server/game/Entities/Player/Player.h b/src/server/game/Entities/Player/Player.h
index dcaa219..1be2e75 100644
--- a/src/server/game/Entities/Player/Player.h
+++ b/src/server/game/Entities/Player/Player.h
@@ -2467,6 +2467,7 @@ class TC_GAME_API Player : public Unit, public GridObject<Player>
         bool   m_SeasonalQuestChanged;
         time_t m_lastDailyQuestTime;
 
+        uint32 m_hostileReferenceCheckTimer;
         uint32 m_drunkTimer;
         uint32 m_weaponChangeTimer;
 
 src/server/game/Entities/Player/Player.cpp | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/src/server/game/Entities/Player/Player.cpp b/src/server/game/Entities/Player/Player.cpp
index 21d1f25..c5f7727 100644
--- a/src/server/game/Entities/Player/Player.cpp
+++ b/src/server/game/Entities/Player/Player.cpp
@@ -380,6 +380,7 @@ Player::Player(WorldSession* session): Unit(true)
     m_MirrorTimerFlags = UNDERWATER_NONE;
     m_MirrorTimerFlagsLast = UNDERWATER_NONE;
     m_isInWater = false;
+	m_hostileReferenceCheckTimer = 0;
     m_drunkTimer = 0;
     m_deathTimer = 0;
     m_deathExpireTime = 0;
@@ -1550,6 +1551,18 @@ void Player::Update(uint32 p_time)
     if (pet && !pet->IsWithinDistInMap(this, GetMap()->GetVisibilityRange()) && !pet->isPossessed())
     //if (pet && !pet->IsWithinDistInMap(this, GetMap()->GetVisibilityDistance()) && (GetCharmGUID() && (pet->GetGUID() != GetCharmGUID())))
         RemovePet(pet, PET_SAVE_NOT_IN_SLOT, true);
+	
+	if (IsAlive())
+    {
+        if (m_hostileReferenceCheckTimer <= p_time)
+        {
+            m_hostileReferenceCheckTimer = 15 * IN_MILLISECONDS;
+            if (!GetMap()->IsDungeon())
+                getHostileRefManager().deleteReferencesOutOfRange(GetVisibilityRange());
+        }
+        else
+            m_hostileReferenceCheckTimer -= p_time;
+    }
 
     //we should execute delayed teleports only for alive(!) players
     //because we don't want player's ghost teleported from graveyard

 src/server/game/Combat/HostileRefManager.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/src/server/game/Combat/HostileRefManager.h b/src/server/game/Combat/HostileRefManager.h
index 859bee3..2f414fb 100644
--- a/src/server/game/Combat/HostileRefManager.h
+++ b/src/server/game/Combat/HostileRefManager.h
@@ -53,6 +53,9 @@ class TC_GAME_API HostileRefManager : public RefManager<Unit, ThreatManager>
         // Remove specific faction references
         void deleteReferencesForFaction(uint32 faction);
 
+        // for combat bugs
+        void deleteReferencesOutOfRange(float range);
+
         HostileReference* getFirst() { return ((HostileReference*) RefManager<Unit, ThreatManager>::getFirst()); }
 
         void updateThreatTables();

 src/server/game/Combat/HostileRefManager.cpp | 20 ++++++++++++++++++++
 1 file changed, 20 insertions(+)

diff --git a/src/server/game/Combat/HostileRefManager.cpp b/src/server/game/Combat/HostileRefManager.cpp
index 965b910..3b3ea24 100644
--- a/src/server/game/Combat/HostileRefManager.cpp
+++ b/src/server/game/Combat/HostileRefManager.cpp
@@ -142,6 +142,26 @@ void HostileRefManager::deleteReferencesForFaction(uint32 faction)
 }
 
 //=================================================
+// delete all references out of specified range
+
+void HostileRefManager::deleteReferencesOutOfRange(float range)
+{
+    HostileReference* ref = getFirst();
+    range = range*range;
+    while (ref)
+    {
+        HostileReference* nextRef = ref->next();
+        Unit* owner = ref->GetSource()->GetOwner();
+        if (!owner->isActiveObject() && owner->GetExactDist2dSq(GetOwner()) > range)
+        {
+            ref->removeReference();
+            delete ref;
+        }
+        ref = nextRef;
+    }
+}
+
+//=================================================
 // delete one reference, defined by Unit
 
 void HostileRefManager::deleteReference(Unit* creature)

 src/server/game/AI/CoreAI/PetAI.cpp | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/src/server/game/AI/CoreAI/PetAI.cpp b/src/server/game/AI/CoreAI/PetAI.cpp
index 9ef589a..971067f 100644
--- a/src/server/game/AI/CoreAI/PetAI.cpp
+++ b/src/server/game/AI/CoreAI/PetAI.cpp
@@ -48,6 +48,11 @@ bool PetAI::_needToStop()
     if (me->IsCharmed() && me->GetVictim() == me->GetCharmer())
         return true;
 
+    // dont allow pets to follow targets far away from owner
+    if (Unit* owner = me->GetCharmerOrOwner())
+        if (owner->GetExactDist(me) >= (owner->GetVisibilityRange()-10.0f))
+            return true;
+
     return !me->IsValidAttackTarget(me->GetVictim());
 }
 
```

</details>

**Comment by sirikfoll** [comment]
Is very likely that all the main issues here are already fixed, any remaining should be due to bad scripting in instances.
If you are still have some issue of this kind, please post here which NPC still bugs.
Make sure to test it on an updated version of the core.

**Comment by Killyana** [comment]
For Jotunheim Proto-Drake there's already an issue for that, and it will probably be fixed by Treeston: Combat/threat system rewrite.

**Comment by Aokromes** [comment]
Test after 9f9507e

**Comment by Treeston** [comment]
This can be assumed fixed by 9f9507e unless more cases come up.
